src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -765,7 +765,146 @@ private ITypeReference GetUnionTypeType(UnionTypeSyntax syntax)
+                        .FeatureIsTemporarilyDisabled("tagged unions"));
+                // return FinalizeDiscriminatedObjectType(syntax, unionMembers, discriminatorDecorator);
+            }
+
+            return TypeHelper.CreateTypeUnion(unionMembers.Select(t => t.Item2));
+        }
+
+        private DecorableSyntax? TryResolveUnionImmediateDecorableSyntax(SyntaxBase? syntaxBase) =>
+            syntaxBase switch
+            {
+                DecorableSyntax decorableSyntax => decorableSyntax,
+                ParenthesizedExpressionSyntax or UnionTypeSyntax or UnionTypeMemberSyntax or NullableTypeSyntax =>
+                    TryResolveUnionImmediateDecorableSyntax(binder.GetParent(syntaxBase)),
+                _ => null
+            };
+
+        private bool IsNullableUnion(UnionTypeSyntax unionTypeSyntax) =>
+            binder.GetParent(unionTypeSyntax) is ParenthesizedExpressionSyntax parenthesizedExpressionSyntax
+            && binder.GetParent(parenthesizedExpressionSyntax) is NullableTypeSyntax;
+
+#pragma warning disable IDE0051
+        private TypeSymbol FinalizeDiscriminatedObjectType(
+#pragma warning restore IDE0051
+            UnionTypeSyntax unionTypeSyntax,
+            IEnumerable<(UnionTypeMemberSyntax syntax, ITypeReference type)> unionMembers,
+            DecoratorSyntax discriminatorDecorator)
+        {
+            var discriminatorPropertyExpr = discriminatorDecorator.Arguments.FirstOrDefault()?.Expression as StringSyntax;
+            var discriminatorPropertyName = discriminatorPropertyExpr?.TryGetLiteralValue();
+
+            if (discriminatorPropertyName == null)
+            {
+                return ErrorType.Empty(); // the decorator validator handles this case
+            }
+
+            var errorDiagnostics = new List<ErrorDiagnostic>();
+            // NOTE(kylealbert): keys are bicep string literals (ex: "'a'" and not "a")
+            var memberDiscriminatorValues = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase); // back end is case insensitive
+            var expandedMemberTypes = new List<ObjectType>();
+
+            foreach (var (memberSyntax, memberType) in unionMembers)
+            {
+                var memberTypeEvaluated = memberType.Type;
+
+                if (memberTypeEvaluated is ObjectType objectType)
+                {
+                    // validate the member has the discriminator property defined and is of the right type
+                    if (!objectType.Properties.TryGetValue(discriminatorPropertyName, out var discriminatorTypeProperty)
+                        || discriminatorTypeProperty.TypeReference.Type is not StringLiteralType discriminatorMemberLiteral
+                        || !discriminatorTypeProperty.Flags.HasFlag(TypePropertyFlags.Required)
+                        || !string.Equals(discriminatorPropertyName, discriminatorTypeProperty.Name, LanguageConstants.IdentifierComparison))
```

Comment:
Isn't this last check just repeating line 824?

Reply:
This does appear to be a duplicate. I did a case-sensitivity pass later and made a mistake here.
Removed in https://github.com/Azure/bicep/pull/11222/commits/b889fea2e5d0e05eff348307313020b7cf24009b